R/aggregate_game_stats_def.R

```diff
@@ -0,0 +1,527 @@
+        tidyselect::contains("sack_"),
+        tidyselect::starts_with("qb_hit_")
+      ),
+      names_to = "desc",
+      names_prefix = "sk_",
+      values_to = "player_id",
+      values_drop_na = TRUE
+    ) %>%
+    dplyr::mutate(
+      n = dplyr::case_when(
+        desc %in% c("half_sack_1_player_id", "half_sack_2_player_id") ~ 0.5,
+        TRUE ~ 1
+      ),
+      desc = stringr::str_remove_all(.data$desc,"_player_id") %>%
+        stringr::str_remove_all("_[0-9]") %>%
+        stringr::str_remove("half_")
+    ) %>%
+    tidyr::pivot_wider(
+      names_from = .data$desc,
+      values_from = c(.data$n,.data$yards_gained),
+      values_fn = sum,
+      values_fill = 0
+    ) %>%
+    dplyr::select(
+      "season",
+      "week",
+      "team",
+      "player_id",
+      "sacks" = "n_sack",
+      "qb_hit" = "n_qb_hit",
+      "sack_yards" = "yards_gained_sack"
+    )
+
+  # Interception and Deflection stats ---------------------------------------------------------
+
+  # get int and def stats
+  int_df <- data %>%
+    dplyr::select(
+      "season", "week","return_yards","team"="defteam",
+      tidyselect::starts_with("interception_"),
+      tidyselect::starts_with("pass_defense_"),
+      -tidyselect::contains("_name")) %>%
+    tidyr::pivot_longer(
+      cols = c(
+        tidyselect::starts_with("interception_"),
+        tidyselect::starts_with("pass_defense_")
+      ),
+      names_to = "desc",
+      names_prefix = "int_",
+      values_to = "db_player_id",
+      values_drop_na = TRUE
+    ) %>%
+    dplyr::mutate(
+      n = 1,
+      desc = stringr::str_remove_all(.data$desc,"_player_id") %>%
+        stringr::str_remove_all("_[0-9]")
+    ) %>%
+    tidyr::pivot_wider(names_from = "desc",
+                       values_from = c("n","return_yards"),
+                       values_fn = sum,
+                       values_fill = 0) %>%
+    dplyr::select(
+      "season",
+      "week",
+      "team",
+      "player_id" = "db_player_id",
+      "int" = "n_interception",
+      "pass_defended" = "n_pass_defense",
+      "int_yards" = "return_yards_interception"
+    )
+
+  # Safety stats -----------------------------------------------------------
+
+  safety_df <- data %>%
+    dplyr::filter(.data$safety==1,!is.na(.data$safety_player_id)) %>%
+    dplyr::select("season","week","team"="defteam","player_id" = "safety_player_id") %>%
+    dplyr::count(.data$season,.data$week,.data$team,.data$player_id, name = "safety")
+
+  # Fumble stats -----------------------------------------------------------
+
+  # get fumble stats for fumbles and own fumble recoveries
+  fumble_df_own <- data %>%
+    dplyr::filter(.data$fumble == 1 | .data$fumble_lost == 1) %>%
+    dplyr::filter(
+      .data$defteam == .data$fumbled_1_team |
+        .data$defteam == .data$fumbled_2_team
+    ) %>%
+    dplyr::mutate(
+      fumbled_1_player_id =
+        ifelse(.data$defteam == .data$fumbled_1_team, .data$fumbled_1_player_id, NA)
+    ) %>%
```

Comment:
We've got a bug here. In order for there to be an own fumble recovery, we need to make sure that the fumble recovery player name matches the fumble player name on the **same fumble**. This current code incorrectly gives Christian Wilkins an own fumble recovery for 9 yards in Week 11 of 2011. (But on the play he recovers an opponent's fumble , takes it 9 yards, and fumbles it himself, but his teammate recovers).

Reply:
You can move this to another issue as a known bug and then propose code to fix it once this is merged.